grpclib/protocol.py

```diff
@@ -223,13 +230,88 @@ def flush(self) -> None:
         if data:
             self._transport.write(data)
 
+    def initialize(self) -> None:
+        if self._config._keepalive_time is not None:
+            self._ping_handle = asyncio.get_event_loop().call_later(
+                self._config._keepalive_time,
+                self._ping
+            )
+
     def close(self) -> None:
         if hasattr(self, '_transport'):
             self._transport.close()
             # remove cyclic references to improve memory usage
             del self._transport
             if hasattr(self._connection, '_frame_dispatch_table'):
                 del self._connection._frame_dispatch_table
+        if self._ping_handle is not None:
+            self._ping_handle.cancel()
+        if self._close_by_ping_handler is not None:
+            self._close_by_ping_handler.cancel()
+
+    def _ping(self) -> None:
+        data = struct.pack('!Q', int(time.monotonic() * 10 ** 6))
+        is_need_send = True
+
+        if not self._config._keepalive_permit_without_calls:
+            count_of_connection = any(
+                s.open for s in self._connection.streams.values()
+            )
+            if count_of_connection == 0:
+                is_need_send = False
+
+        _keepalive_time = cast(float, self._config._keepalive_time)
+        current_time = time.monotonic()
+        if current_time - self._last_received() < _keepalive_time:
```

Comment:
You don't have to perform a next check if a previous check resulted in `is_need_send = False`

Reply:
```python
    def _is_ping_allowed(self):
        if not a or not b:
            return False
        if not c or not d:
            return False
        return True

    def _ping(self):
        if self._is_ping_allowed():
            data = struct.pack(...)
            self._connection.ping(data)
        ...call_later(..., self._ping)
```

Reply:
done